test/framework/deployment_helpers.go

```diff
@@ -132,26 +177,41 @@ func WatchDeploymentLogs(ctx context.Context, input WatchDeploymentLogsInput) {
 
 	selector, err := metav1.LabelSelectorAsMap(deployment.Spec.Selector)
 	Expect(err).NotTo(HaveOccurred(), "Failed to Pods selector for deployment %s", klog.KObj(input.Deployment))
+	watchPodLogs(ctx, watchPodLogsInput{
+		GetLister:      input.GetLister,
+		ClientSet:      input.ClientSet,
+		Namespace:      deployment.Namespace,
+		DeploymentName: deployment.Name,
+		Labels:         selector,
+		LogPath:        input.LogPath,
+	})
+}
 
+// watchPodLogs streams logs for all containers for all pods belonging to a deployment with the given label. Each container's logs are streamed
+// in a separate goroutine so they can all be streamed concurrently. This only causes a test failure if there are errors
+// retrieving the deployment, its pods, or setting up a log file. If there is an error with the log streaming itself,
+// that does not cause the test to fail.
+func watchPodLogs(ctx context.Context, input watchPodLogsInput) {
 	pods := &corev1.PodList{}
-	Expect(input.GetLister.List(ctx, pods, client.InNamespace(input.Deployment.Namespace), client.MatchingLabels(selector))).To(Succeed(), "Failed to list Pods for deployment %s", klog.KObj(input.Deployment))
+	Expect(input.GetLister.List(ctx, pods, client.InNamespace(input.Namespace), client.MatchingLabels(input.Labels))).To(Succeed(), "Failed to list Pods for deployment %s", input.DeploymentName)
 
 	for _, pod := range pods.Items {
-		for _, container := range deployment.Spec.Template.Spec.Containers {
-			log.Logf("Creating log watcher for controller %s, pod %s, container %s", klog.KObj(input.Deployment), pod.Name, container.Name)
+		for _, container := range pod.Spec.Containers {
+			log.Logf("Creating log watcher for controller %s, pod %s, container %s", input.DeploymentName, pod.Name, container.Name)
 
 			// Create log metadata file.
-			logMetadataFile := filepath.Clean(path.Join(input.LogPath, input.Deployment.Name, pod.Name, container.Name+"-log-metadata.json"))
+			logMetadataFile := filepath.Clean(path.Join(input.LogPath, pod.Name, pod.Name, container.Name+"-log-metadata.json"))
 			Expect(os.MkdirAll(filepath.Dir(logMetadataFile), 0750)).To(Succeed())
 
 			metadata := logMetadata{
-				Job:       input.Deployment.Namespace + "/" + input.Deployment.Name,
-				Namespace: input.Deployment.Namespace,
-				App:       input.Deployment.Name,
+				Job:       input.Namespace + "/" + pod.Name,
+				Namespace: input.Namespace,
+				App:       input.DeploymentName,
 				Pod:       pod.Name,
 				Container: container.Name,
 				NodeName:  pod.Spec.NodeName,
 				Stream:    "stderr",
+				Labels:    input.Labels,
```

Comment:
I'm not sure we should really propagate the labels here

Reply:
I would be careful given that I have no idea if Kubernetes labels will break Loki / if we really want to propagate them

Reply:
I have removed the labels. I added them because I thought if we are watching logs by LabelSelector it would be nice to see them in collected logs.

Reply:
Would be if it works. I'm just not sure if it does. E.g. Prometheus has problems if you have two labels for which the only difference is `-` vs `_`. I'm not sure which rules Loki has for labels and how they match with what Kubernetes allows for its labels